dhis-2/dhis-test-integration/src/test/java/org/hisp/dhis/program/notification/ProgramNotificationInstanceServiceTest.java

```diff
@@ -0,0 +1,141 @@
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package org.hisp.dhis.program.notification;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+import java.time.Instant;
+import java.util.Date;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Stream;
+import org.hisp.dhis.common.IdentifiableObjectManager;
+import org.hisp.dhis.common.UID;
+import org.hisp.dhis.organisationunit.OrganisationUnit;
+import org.hisp.dhis.program.Enrollment;
+import org.hisp.dhis.program.Program;
+import org.hisp.dhis.security.Authorities;
+import org.hisp.dhis.test.integration.PostgresIntegrationTestBase;
+import org.hisp.dhis.trackedentity.TrackedEntity;
+import org.hisp.dhis.trackedentity.TrackedEntityType;
+import org.hisp.dhis.tracker.imports.job.NotificationSender;
+import org.hisp.dhis.tracker.imports.programrule.engine.Notification;
+import org.hisp.dhis.user.User;
+import org.hisp.dhis.user.UserGroup;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.transaction.annotation.Transactional;
+
+/**
+ * @author Zubair Asghar
+ */
+@Transactional
+class ProgramNotificationInstanceServiceTest extends PostgresIntegrationTestBase {
+  private static final int TEST_USER_COUNT = 60;
+  private static final Date NOW = Date.from(Instant.now());
+  private static final int EXPECTED_NOTIFICATIONS = 20;
+
+  @Autowired private ProgramNotificationInstanceService programNotificationInstanceService;
+  @Autowired private IdentifiableObjectManager manager;
+  @Autowired private NotificationSender notificationSender;
+
+  private OrganisationUnit organisationUnit;
+  private UserGroup userGroup;
+
+  @BeforeEach
+  void setUp() {
+    organisationUnit = createOrganisationUnit('O');
+    manager.save(organisationUnit);
+    Program program = createProgram('P', Set.of(), organisationUnit);
+    manager.save(program);
+    userGroup = createUserGroup('U', new HashSet<>());
+    manager.save(userGroup);
+    createAndAddUsersToUserGroupAndOrgUnit();
+
+    TrackedEntityType trackedEntityType = createTrackedEntityType('P');
+    manager.save(trackedEntityType);
+    TrackedEntity trackedEntity = createTrackedEntity(organisationUnit, trackedEntityType);
+    manager.save(trackedEntity);
+    Enrollment enrollment = createEnrollment(program, trackedEntity, organisationUnit);
+    manager.save(enrollment);
+
+    getNotifications().forEach(n -> notificationSender.send(n, enrollment));
+  }
+
+  @Test
+  @DisplayName("Should fetch scheduled notifications within timeout")
+  void testScheduledNotificationShouldBeTriggeredBeforeTimeout() {
```

Comment:
Name it `shouldReturnProgramNotificationInstancesForGivenDate` as that is what you are testing. The fact that this would have timed out in the past is secondary. There is no need to mention that in our code. One can get that context from the issue/PR.